### PR TITLE
Implemented Book CRUD

### LINKED ISSUE #7: Implement CRUD for Book
System operations that need to be implemented:

- Book search
- Add book 
- Update book 
- Delete book

### PR DESCRIPTION
**Pull request by VladHolobyn**
Closes #7

### CODE REVIEW

File: src/main/java/com/pnudev/librarysystem/entity/Book.java [hunk 1]
```diff
@@ -44,12 +49,32 @@ public class Book {
...
+    @JoinTable(
+            name = "book-author",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "author_id")
+    )
     private List<Author> authors;
 
-    @ManyToMany(mappedBy = "books")
+    @ElementCollection
+    @CollectionTable(name="book-author", joinColumns=@JoinColumn(name="book_id"))
```
Comment (justme1135) [inline src/main/java/com/pnudev/librarysystem/entity/Book.java]: table namings are written `like_that`

File: src/main/java/com/pnudev/librarysystem/service/AuthorService.java [hunk 2]
```diff
@@ -15,6 +15,7 @@
 import org.springframework.stereotype.Service;
 
 
+
```
Comment (justme1135) [inline src/main/java/com/pnudev/librarysystem/service/AuthorService.java]: remove

File: src/main/java/com/pnudev/librarysystem/mapper/BookMapper.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+@Mapper(componentModel = "spring",
+uses = {CategoryMapper.class, AuthorMapper.class})
+public interface BookMapper {
+
+    @Mapping(target = "coverImage", source = "coverImageName")
+    Book toEntity(RequestBookDTO requestBookDTO);
+
+    @Mapping(
+            target = "coverImageUrl",
+            expression = "java(\"/books/\" + book.getId() + \"/image\")"
```
Comment (justme1135) [inline src/main/java/com/pnudev/librarysystem/mapper/BookMapper.java]: refactor using `.formatted`

File: src/main/java/com/pnudev/librarysystem/util/SpecificationUtils.java [hunk 4]
```diff
@@ -9,4 +9,10 @@
     public static <T> Specification<T> fieldContainsIgnoreCase(String field, String value){
```
Comment (justme1135) [inline src/main/java/com/pnudev/librarysystem/util/SpecificationUtils.java]: If you are going to fully "generify" these utilities, you might as well make them an abstract class and then extend it where needed
Also it should be named something like `AbstractSpecificationBuilder`

**Review by justme1135 — CHANGES_REQUESTED** [review]

File: src/main/java/com/pnudev/librarysystem/service/BookService.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    public Long createBook(RequestBookDTO requestBookDTO)  {
+        Book book = bookMapper.toEntity(requestBookDTO);
+        Book savedBook = bookRepository.save(book);
+        return savedBook.getId();
+    }
+
+    public BookDTO updateBook(Long id, RequestBookDTO requestBookDTO) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book with id: %d not found".formatted(id)));
```
Comment (getmane) [inline src/main/java/com/pnudev/librarysystem/service/BookService.java]: I see a lot of throws with the exact same message format. 
You can extract this as a constant and just pass on the id when needed

File: src/main/java/com/pnudev/librarysystem/service/BookService.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        return bookMapper.toDTO(book);
+    }
+
+    public String uploadCoverImage(MultipartFile file) {
+        if(file.isEmpty()) {
+            throw new EmptyFileException("File cannot be empty");
+        }
+
+        String originalFileName = file.getOriginalFilename();
+        String fileExtension = originalFileName.substring(originalFileName.lastIndexOf(".")+1);
```
Comment (getmane) [inline src/main/java/com/pnudev/librarysystem/service/BookService.java]: StringUtils.substringAfterLast for the file extension

File: src/main/java/com/pnudev/librarysystem/service/BookService.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+        if (!imageTypes.contains(fileExtension)) {
+            throw new FileWrongTypeException(imageTypes);
+        }
+
+        Path uploadPath= Paths.get(uploadDir);
```
Comment (getmane) [inline src/main/java/com/pnudev/librarysystem/service/BookService.java]: Whitespace

File: src/main/java/com/pnudev/librarysystem/service/BookService.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            } catch (IOException e) {
+                throw new IOErrorInFileException("Cannot create directory: " + e.getMessage());
+            }
+        }
+
+        String filename = UUID.randomUUID() + "." + fileExtension;
+        Path imagePath = uploadPath.resolve(filename);
+
+        try {
+            Files.copy(file.getInputStream(),imagePath, StandardCopyOption.REPLACE_EXISTING);
```
Comment (getmane) [inline src/main/java/com/pnudev/librarysystem/service/BookService.java]: Whitespace (you can do auto-formatting in the intellij, ctrl-alt-L)

File: src/main/java/com/pnudev/librarysystem/service/BookService.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        bookMapper.updateBookFromRequestDTO(requestBookDTO, book);
+
+        return bookMapper.toDTO(bookRepository.save(book));
+    }
+
+    public void deleteBookById(Long id) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book with id: %d not found".formatted(id)));
+
+        if(!book.getBorrowings().isEmpty()) {
```
Comment (getmane) [inline src/main/java/com/pnudev/librarysystem/service/BookService.java]: Whitespace after if

File: src/main/java/com/pnudev/librarysystem/service/CategoryService.java [hunk 10]
```diff
@@ -11,6 +11,7 @@
 import org.springframework.data.domain.PageRequest;
 import org.springframework.stereotype.Service;
 
+
```
Comment (getmane) [inline src/main/java/com/pnudev/librarysystem/service/CategoryService.java]: Was this newline needed?

File: src/main/java/com/pnudev/librarysystem/service/BookService.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    public BookDTO getBookById(Long id) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book with id: %d not found".formatted(id)));
+        return bookMapper.toDTO(book);
+    }
+
+    public String uploadCoverImage(MultipartFile file) {
+        if(file.isEmpty()) {
```
Comment (getmane) [inline src/main/java/com/pnudev/librarysystem/service/BookService.java]: Whitespace

File: src/main/java/com/pnudev/librarysystem/entity/Book.java (hunk 1 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/entity/Book.java]: Renamed

File: src/main/java/com/pnudev/librarysystem/service/AuthorService.java (hunk 2 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/service/AuthorService.java]: Removed

File: src/main/java/com/pnudev/librarysystem/mapper/BookMapper.java (hunk 3 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/mapper/BookMapper.java]: Refactored

File: src/main/java/com/pnudev/librarysystem/util/SpecificationUtils.java (hunk 4 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/util/SpecificationUtils.java]: Done, if I understood you correctly

File: src/main/java/com/pnudev/librarysystem/service/BookService.java (hunk 5 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/service/BookService.java]: Done

File: src/main/java/com/pnudev/librarysystem/service/BookService.java (hunk 6 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/service/BookService.java]: Done

File: src/main/java/com/pnudev/librarysystem/service/BookService.java (hunk 7 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/service/BookService.java]: Formatted

File: src/main/java/com/pnudev/librarysystem/service/BookService.java (hunk 8 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/service/BookService.java]: Formatted

File: src/main/java/com/pnudev/librarysystem/service/BookService.java (hunk 9 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/service/BookService.java]: Formatted

File: src/main/java/com/pnudev/librarysystem/service/CategoryService.java (hunk 10 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/service/CategoryService.java]: No)

File: src/main/java/com/pnudev/librarysystem/service/BookService.java (hunk 11 above)
Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/service/BookService.java]: Formatted

File: src/main/java/com/pnudev/librarysystem/service/CategoryService.java (hunk 10 above)
Comment (getmane) [reply src/main/java/com/pnudev/librarysystem/service/CategoryService.java]: So, remove it?)

Comment (VladHolobyn) [reply src/main/java/com/pnudev/librarysystem/service/CategoryService.java]: OK

**Review by getmane — APPROVED** [review]

**Pull request merged**